rfd/0088-passwordless-windows.md

```diff
@@ -0,0 +1,216 @@
+
+```shell
+$ tsh mfa add
+> Choose device type [TOTP, WEBAUTHN]: webauthn
+> Enter device name: win-hello
+> Tap any *registered* security key or enter a code from a *registered* OTP device: <taps>
+<system shows webauthn prompt>
+> MFA device "win-hello" added.
+```
+
+The way to specify if windows hello or FIDO device should be used is via
+mfa-mode: `--mfa-mode={auto,platform,cross-platform}`. Auto defaults to platform
+in Windows Hello.
+
+`tsh` performs the key registration process, setting all parameters for a
+passwordless / resident key via `WebAuthNAuthenticatorMakeCredential` from
+`webauthn.dll`.
+
+<!--
+TODO(codingllama): should we add here some code samples? same for authenticaiton
+-->
+
+In case of a registration failure, `tsh` must do its best to delete the
+created-but-not-registered credential. Credential removal is only possible
+using webauthn API version 4+.
+
+### Authentication
+
+Authentication depends on server settings (otp, webauthn, passwordless).
+
+`tsh` performs the authentication process, setting all parameters for a
+passwordless / resident key via `WebAuthNAuthenticatorGetAssertion` from
+`webauthn.dll`.
+
+The way to specify if Windows Hello or FIDO device should be used is via
+mfa-mode: `tsh --mfa-mode={auto,platform,cross-platform}` - choose whether to
+use platform or cross-platform MFA. Auto defaults to platform in Windows Hello.
+
+### Detecting Windows Hello support
+
+Detecting Windows Hello and FIDO devices support is important so `tsh` may
+enable/disable related features as appropriate.
+
+Apart from Go build tags, which are a rather coarse detection mechanism, we can
+
+* Verify if `webauthn.dll` is present on system and it's version (support v1-v4)
+* Verify device supports for user-verifying platform authenticator via API call
+  `WebAuthNIsUserVerifyingPlatformAuthenticatorAvailable`
+* If device is not supporting user-verifying platform authenticator, fallback to
+  FIDO devices should be possible
+
+### `tsh` support commands
+
+The following support commands are added to `tsh` as hidden subcommands. They
+are useful to diagnose and manage certain aspects of Windows Hello support.
+
+The commands are only available on windows builds.
+
+`tsh winhello diag` - prints diagnostics about Windows Hello support (for
+example, what API version is `webauthn.dll` etc.
+
+```shell
+$ tsh winhello diag  # diag output subject to change
+> Windows version: 10
+> Signed: yes
+> Webauthn.dll version: 3
+> Supports platform user-verification: yes
+```
+
+### webauthn.dll and Teleport
+
+At the time of writing this rfd, Go wrapper around `webauthn.dll` does not
+exists. New open source repository will be created where teleport will maintain
+unofficial Go client for `webauthn.dll`.
+
+### Executable sign
+
+Although Windows allow usage of not-signed executables, it results in poor UX.
+There is warning about _using not verified app_. This can be easily fixed by
+signing the executable.
+
+### Security
```

Comment:
Let's talk a bit about how Hello works, for those unversed (including myself). For example:

* How does Hello key storage works? Is it TPM-based?
* How are keys scoped?
* Can any application access the keys?
* Hoes does signing / not-signing the application changes the behavior above?

Reply:
> How does Hello key storage works? Is it TPM-based?
It will be TPM based is device supports TPM 2.0. If not, it will just software (https://docs.microsoft.com/en-us/windows/security/identity-protection/hello-for-business/hello-faq#how-are-keys-protected). Should we try to detect if device supports TPM by ourselves?

>How are keys scoped?
Keys are scoped by rpid

>Can any application access the keys?
Yes, I was able to use win-hello key inside Chrome which was created from my simple Go application. You just need to specify same rpid.

>Hoes does signing / not-signing the application changes the behavior above?
It does not change anyhow behaviour. If you use signed app to create keys, you can still access it from other apps. Same for unsigned app.

>Let's also be explicit about whether a credential added via tsh is expected to work in the web UI to set expectations. This should also end up being documented in bold print so users don't get confused.
It will work. I will add it into RFD.

Reply:
> > How does Hello key storage works? Is it TPM-based?
> 
> It will be TPM based is device supports TPM 2.0. If not, it will just software (https://docs.microsoft.com/en-us/windows/security/identity-protection/hello-for-business/hello-faq#how-are-keys-protected). Should we try to detect if device supports TPM by ourselves?

Windows admins can enforce TPM, that seems sufficient to me. Let's just make sure to recommend TPM-backed Hello in public docs.

> > How are keys scoped?
> 
> Keys are scoped by rpid
> 
> > Can any application access the keys?
> 
> Yes, I was able to use win-hello key inside Chrome which was created from my simple Go application. You just need to specify same rpid.

Key usage always require some sort of user verification, right (biometric, PIN, etc)? 

I think it's worth mentioning this in the Security section, as it differs from other systems, but as long as some verification happens it's not that different from a hardware fob.

Reply:
> Key usage always require some sort of user verification, right (biometric, PIN, etc)?

> I think it's worth mentioning this in the Security section, as it differs from other systems, but as long as some verification happens it's not that different from a hardware fob.


Yes key usage of windows hello requires biometrics or pin, example: 
<img width="342" alt="0088-login-choice-hello" src="https://user-images.githubusercontent.com/14020794/190571181-402aa76b-2b58-4c53-9881-01b9479e5f3b.png">

I will update security section 👌